File: java/src/ch1/ex116.java
package ch1;

public class ex116 {
    public static void main(String[] args) {
        System.out.println(expt(2, 0));
        System.out.println(expt(2, 4));
        System.out.println(expt(2, 9));
        System.out.println(expt(3, 4));
        System.out.println(expt(3, 5));
        System.out.println("done");
    }

    static int expt(int b, int p) {
        return expt_invariant(b, p, 1);
    }

    static int expt_invariant(int b, int p, int a) {
        if (p == 0) return a;
        if (even(p)) return expt_invariant(sq(b), p / 2, a);
        return expt_invariant(b, p - 1, a * b);
    }

    static int sq(int x) {
        return x * x;
    }

    static boolean even(int x) {
        return x % 2 == 0;
    }
}

File: javascript/ch1/Ex116.js
// Load with: var ExXXX = require('./ExXXX')
// Run with: ExXXX.test()

sq = function (x) {
    return x * x;
};

even = function (x) {
    return x % 2 == 0;
};

expt_invariant = function (b, p, a) {
    if (p == 0) return a;
    if (even(p)) return expt_invariant(sq(b), p / 2, a);
    return expt_invariant(b, p - 1, a * b);
};

exports.expt = function (b, p) {
    return expt_invariant(b, p, 1);
};

exports.test = function() {
    console.log(exports.expt(2,0));
    console.log(exports.expt(2,1));
    console.log(exports.expt(2,2));
    console.log(exports.expt(2,8));
    console.log(exports.expt(2,9));
    console.log(exports.expt(3, 4));
};